morango/models/utils.py

```diff
@@ -0,0 +1,197 @@
+
+
+def _query_wmic(namespace, key):
+    try:
+        result = (
+            subprocess.check_output("wmic {} get {}".format(namespace, key))
+            .decode()
+            .split()[-1]
+        )
+
+        if "-" in result:
+            return result
+    except:  # noqa: E722
+        pass
+
+
+def _get_macos_uuid():
+    try:
+        command = "ioreg -rd1 -c IOPlatformExpertDevice | grep -E '(UUID)'"
+        result = subprocess.check_output(command, shell=True)
+        return result.decode().split('"')[-2]
+    except:  # noqa: E722
+        pass
+
+
+def _get_android_uuid():
+    try:
+        for propname in ["ro.serialno", "ril.serialnumber", "ro.boot.serialno"]:
+            output = subprocess.check_output(["getprop", propname]).decode().strip()
+            if output:
+                return output
+    except:  # noqa: E722
+        pass
+
+
+def _do_salted_hash(value):
+    if not value:
+        return ""
+    if isinstance(value, int):
+        value = str(value)
+    try:
+        value = value.encode()
+    except:  # noqa: E722
+        pass
+    value += "::b88281f3-302c-4def-bb87-043f681e4183".encode()
+    value = value.lower()
+    return hashlib.sha1(value).hexdigest()[:20]
+
+
+def get_0_5_system_id():
+    """
+    See https://github.com/learningequality/morango/issues/79#issuecomment-590584559
+    """
+
+    # check whether envvar was set, and use that if available
+    system_id = os.environ.get("MORANGO_SYSTEM_ID")
+    if system_id and len(system_id.strip()) >= 3:
+        # skip to the end of the elif's to return
+        pass
+
+    # Windows
+    elif sys.platform == "win32":
+
+        # try to get the system serial number, if available
+        system_id = _query_wmic("csproduct", "UUID")
+        # if UUID consists only of "F" digits, it's not usable
+        if system_id and not system_id.replace("F", "").replace("-", "").strip():
+            system_id = None
+
+        # otherwise, try to get the serial number of the disk drive
+        if not system_id:
+            system_id = _query_wmic("DISKDRIVE", "SerialNumber")
+
+    # MacOS
+    elif sys.platform == "darwin":
+        system_id = _get_macos_uuid()
+
+    # Android
+    elif "ANDROID_ARGUMENT" in os.environ:
+        system_id = _get_android_uuid()
+
+    # Linux
+    elif sys.platform.startswith("linux"):
+        if os.path.isfile("/etc/machine-id"):
+            with open("/etc/machine-id") as f:
+                system_id = f.read().strip()
+
+    return _do_salted_hash(system_id)
+
+
+def _device_sort_key(iface):
+    """
+    Sort interfaces by device name to give preference to interfaces
+    that are more likely to be stable (not change/be removed).
+    """
+    dev = (iface.get("device") or "").lower()
+    if dev.startswith("eth") or dev.startswith("en"):
+        return "0" + dev
+    if dev.startswith("wl"):
+        return "1" + dev
+    if dev.startswith("e") or dev.startswith("w"):
+        return "2" + dev
```

Comment:
Would it be helpful to exclude bridged connections entirely `dev.startswith("br")`? I know with specific Docker network configurations Kolibri would fail to start because of something to do with the mac addresses.

Reply:
I also noticed a vbox*** adapter that had a MAC address, while testing (as I had a virtualbox running in parallel). Don't want to include those either. Wasn't sure I could come up with an exhaustive list of "non-real" interface names, though... especially in a cross-platform way. Any ideas?

Reply:
I'm adding something to check flags in the MAC address for whether it's "local", vs a registered physical address, based on this:
https://en.wikipedia.org/wiki/MAC_address#Universal_vs._local

That seems to exclude at least the vbox case, and should prevent having it pick up on a transient, generated MAC, which would be bad for ID stability.

Reply:
Cool. The exact issue I recall is when you run a docker container with host networking mode, even if it wasn't running, it would cause issues because it was actually the same MAC address as the real adapter. I don't think that should necessarily change what you're doing though.

Reply:
Urgh, yeah, looks as if this was never resolved upstream in ifcfg:
https://github.com/ftao/python-ifcfg/issues/34

The "fix" in Kolibri, apparently by me, was to just... ignore it:
https://github.com/learningequality/kolibri/pull/5914/files

:sigh:

Reply:
I was unable to replicate that issue, and we at least handle it gracefully here, so I don't think it's a blocker at the moment.